File: algorithms/python/arithmeticSlices/arithmeticSlices.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/arithmetic-slices/
# Author: Miao Zhang
# Date:   2021-02-09

class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        n = len(A)
        dp = [0] * n
        for i in range(2, n):
            if A[i - 1] - A[i - 2] == A[i] - A[i - 1]:
                dp[i] = dp[i - 1] + 1
        return sum(dp)

File: algorithms/cpp/arithmeticSlices/arithmeticSlices.cpp
// Source: https://leetcode.com/problems/arithmetic-slices/
// Author: Miao Zhang
// Date:   2021-02-09

class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        int res = 0;
        int cur = 0;
        for (int i = 2; i < A.size(); i++) {
            if (A[i - 1] - A[i - 2] == A[i] - A[i - 1]) {
                res += ++cur;
            } else {
                cur = 0;
            }
        }
        return res;
    }
};
